scripts/test.py

```diff
@@ -0,0 +1,212 @@
+
+    test_permutations = filter_test_permutations([args.test_pattern], [args.blacklist_pattern])
+    if len(test_permutations) == 0:
+        # TODO: support direct test dispatching
+        if args.test_pattern:
+            fail('no tests were selected -- whitelist pattern did not match any known tests')
+        else:
+            fail('no tests were selected -- blacklist is too restrictive')
+
+    print('Preparing to run the following tests:')
+    for (profile, tests) in test_permutations.items():
+        print('- %s:' % profile)
+        for test in tests:
+            print('  - %s' % test)
+
+    timestamp = int(time.time())
+    print('======================================')
+    print('=============', timestamp, '=============')
+    print('======================================')
+
+    log_dir = os.path.join('test_logs', str(timestamp))
+    def make_log_dir():
+        if os.path.exists(log_dir):
+            fail('test log directory already exists -- how???')
+        os.makedirs(log_dir)
+    wet('make new directory: ' + log_dir, make_log_dir)
+
+    for profile in test_permutations.keys():
+        profile_dir = os.path.join(log_dir, profile)
+        wet('make directory: ' + profile_dir, lambda: os.mkdir(profile_dir))
+
+        print('- %s:' % profile)
+        build_log = os.path.join(profile_dir, 'build.log')
+        run_cmd(
+            'dune build --display=progress --profile=%s %s 2> %s'
+                % (profile, coda_exe_path, build_log),
+            lambda: fail_with_log('building %s failed' % profile, build_log)
+        )
+
+        for test in test_permutations[profile]:
+            print('  - %s' % test)
+            log = os.path.join(profile_dir, '%s.log' % test)
+            run_cmd(
+                'set -o pipefail && %s integration-test %s 2>&1 | tee \'%s\' | ./scripts/jqproc.sh -f \'%s\''
+                    % (coda_exe, test, log, jq_filter),
+                lambda: fail('test "%s:%s" failed' % (profile, test))
+            )
+
+    print('all tests ran successfully')
+
+def render(args):
+    circle_ci_conf_dir = os.path.dirname(args.jinja_file)
+    jinja_file_basename = os.path.basename(args.jinja_file)
+    (output_file, ext) = os.path.splitext(args.jinja_file)
+    assert ext == '.jinja'
+
+    test_permutations = filter_test_permutations(['*'], ci_blacklist)
+
+    env = jinja2.Environment(loader=jinja2.FileSystemLoader(circle_ci_conf_dir))
+    template = env.get_template(jinja_file_basename)
+    rendered = template.render(
+        build_artifact_profiles=build_artifact_profiles,
+        unit_test_profiles=unit_test_profiles,
+        test_permutations=test_permutations
+    )
+
+    output_file = open(output_file, 'w')
+    output_file.write(rendered)
+    output_file.close()
+
+def main():
+    actions = {
+        'run': run,
+        'render': render
+    }
+
+    root_parser = argparse.ArgumentParser(description='Coda integration test runner/configurator')
```

Comment:
The help strings aren't very useful. A brief note about the filter language would be useful. Additionally, a command listing all the possibilities for both halves of the pattern would be handy - so you don't have to read the source to figure out the ways this program can be used.